traffic_ops/traffic_ops_golang/api/shared_handlers.go

```diff
@@ -367,34 +369,123 @@ func CreateHandler(creator Creator) http.HandlerFunc {
 		obj := reflect.New(objectType).Interface().(Creator)
 		obj.SetInfo(inf)
 
-		err := decodeAndValidateRequestBody(r, obj)
-		if err != nil {
-			HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, err, nil)
-			return
-		}
+		if c, ok := obj.(AllowMultipleCreates); ok && c.AllowMultipleCreates() {
+			data, err := ioutil.ReadAll(r.Body)
+			if err != nil {
+				HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, err, nil)
+				return
+			}
 
-		if t, ok := obj.(Tenantable); ok {
-			authorized, err := t.IsTenantAuthorized(inf.User)
+			objSlice, err := parseMultipleCreates(data, objectType, inf)
 			if err != nil {
-				HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("checking tenant authorized: "+err.Error()))
+				HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, err)
 				return
 			}
-			if !authorized {
-				HandleErr(w, r, inf.Tx.Tx, http.StatusForbidden, errors.New("not authorized on this tenant"), nil)
+
+			for _, objElemInt := range objSlice {
+				objElem := reflect.ValueOf(objElemInt).Interface().(Creator)
+
+				err = objElem.Validate()
+				if err != nil {
+					HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, err, nil)
+					return
+				}
+
+				if t, ok := objElem.(Tenantable); ok {
+					authorized, err := t.IsTenantAuthorized(inf.User)
+					if err != nil {
+						HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("checking tenant authorized: "+err.Error()))
+						return
+					}
+					if !authorized {
+						HandleErr(w, r, inf.Tx.Tx, http.StatusForbidden, errors.New("not authorized on this tenant"), nil)
+						return
+					}
+				}
+
+				userErr, sysErr, errCode = objElem.Create()
+				if userErr != nil || sysErr != nil {
+					HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+					return
+				}
+
+				if err = CreateChangeLog(ApiChange, Created, objElem, inf.User, inf.Tx.Tx); err != nil {
+					HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, tc.DBError, errors.New("inserting changelog: "+err.Error()))
+					return
+				}
+			}
+			if len(objSlice) == 1 {
+				WriteRespAlertObj(w, r, tc.SuccessLevel, objSlice[0].GetType()+" was created.", objSlice[0])
+			} else {
+				WriteRespAlertObj(w, r, tc.SuccessLevel, objSlice[0].GetType()+"s were created.", objSlice)
```

Comment:
This line makes the assumption that there are 1 or more objects, but there is no previous check to ensure this. Attempting to POST a 0-length array causes this line to error out:

```shell
docker-compose exec mid sh -c 'topost parameters "[]" --insecure --to-user=$TO_ADMIN_USER --to-password=$TO_ADMIN_PASSWORD --to-url=https://$TO_HOST'
```

Response (500):
```json
{"alerts":[{"text":"Internal Server Error","level":"error"}]}
```

Reply:
done